File: beryl/hints.py
import json

from PySide6.QtCore import QObject

# Hint-mode session state. The KeyController consumes every key while in hint
# mode and feeds it here; each keystroke round-trips to hints.js in the
# isolated world (~1-2ms). Uniform-length prefix-free labels mean a unique
# prefix identifies its target — we activate as soon as exactly one remains.


class Hints(QObject):
    def __init__(self, cfg, api, tabs, parent=None):
        super().__init__(parent)
        self._cfg = cfg
        self._api = api
        self._tabs = tabs
        self._keys = None       # set late (keys needs hints, hints needs keys)
        self._typed = ""
        self._new_tab = False

    def set_keys(self, keys):
        self._keys = keys

    def start(self, new_tab):
        self._typed = ""
        self._new_tab = new_tab
        alphabet = json.dumps(self._cfg.get("hint_chars", "asdfghjkl"))
        self._api.js(f"__beryl.hints.show({alphabet})", self._shown, world=1)

    def _shown(self, n):
        if not n:
            self._api.toast.emit("no hints", True)
            return
        self._keys.set_mode("hint")

    def key(self, ks):
        """Called for every key while in hint mode; everything is consumed."""
        if ks == "<Esc>":
            self.cancel()
        elif ks == "<BS>":
            self._typed = self._typed[:-1]
            self._filter()
        elif len(ks) == 1 and ks in self._cfg.get("hint_chars", "asdfghjkl"):
            self._typed += ks
            self._filter()
        # anything else: swallowed silently (scrolling etc cancels via Esc)

    def _filter(self):
        typed = json.dumps(self._typed)
        self._api.js(f"__beryl.hints.filter({typed})", self._filtered, world=1)

    def _filtered(self, live):
        if live == 0 and self._typed:
            self._typed = self._typed[:-1]     # dead key — pretend it didn't happen
            self._filter()
        elif live == 1:
            typed = json.dumps(self._typed)
            new_tab = "true" if self._new_tab else "false"
            self._api.js(f"__beryl.hints.activate({typed}, {new_tab})",
                         self._activated, world=1)

    def _activated(self, res):
        self._keys.set_mode("normal")
        if not isinstance(res, dict):
            return
        if res.get("open"):
            self._tabs.newTab(res["open"], True)   # F = background tab, vimium-style
        # focused → editable.js flips to insert on its own; clicked → done

    def cancel(self):
        self._api.js("__beryl.hints.clear()", world=1)
        self._keys.set_mode("normal")

File: beryl/js/hints.js
// beryl: link hints. Runs in an ISOLATED world (ApplicationWorld) — the page
// can't see or clobber __beryl, and no webchannel is needed because Python
// drives every exchange through runJavaScript round-trips. Labels are
// fixed-position divs inside a closed shadow root, so page CSS can't touch
// them and `* { all: unset }` pages can't break them.
var __beryl = __beryl || {};
__beryl.hints = (function () {
    "use strict";
    var host = null, items = [];

    function clickables() {
        var sel = 'a[href], button, input:not([type=hidden]), textarea, select,' +
                  ' summary, [onclick], [role=link], [role=button], [role=tab],' +
                  ' [role=menuitem], [role=checkbox], [contenteditable=""],' +
                  ' [contenteditable=true], [tabindex]:not([tabindex="-1"])';
        var seen = new Set(), out = [];
        document.querySelectorAll(sel).forEach(function (el) {
            if (seen.has(el)) return;
            seen.add(el);
            var r = el.getBoundingClientRect();
            if (r.width < 2 || r.height < 2) return;
            if (r.bottom < 0 || r.top > innerHeight || r.right < 0 || r.left > innerWidth) return;
            var cs = getComputedStyle(el);
            if (cs.visibility === "hidden" || cs.display === "none" || cs.opacity === "0") return;
            out.push({ el: el, rect: r });
        });
        return out;
    }

    // uniform-length base-k labels — prefix-free by construction
    function labels(n, alphabet) {
        var k = alphabet.length;
        var len = Math.max(1, Math.ceil(Math.log(n) / Math.log(k)));
        var out = [];
        for (var i = 0; i < n; i++) {
            var s = "", x = i;
            for (var j = 0; j < len; j++) { s = alphabet[x % k] + s; x = (x / k) | 0; }
            out.push(s);
        }
        return out;
    }

    function editable(el) {
        if (el.isContentEditable) return true;
        var t = el.tagName;
        if (t === "TEXTAREA" || t === "SELECT") return true;
        if (t === "INPUT")
            return !/^(button|checkbox|radio|submit|reset|file|image|range|color|hidden)$/
                .test((el.type || "text").toLowerCase());
        return false;
    }

    function show(alphabet) {
        clear();
        var els = clickables();
        if (els.length === 0) return 0;
        host = document.createElement("beryl-hints");
        var root = host.attachShadow({ mode: "closed" });
        var box = document.createElement("div");
        box.style.cssText = "position:fixed;inset:0;z-index:2147483647;pointer-events:none;";
        var labs = labels(els.length, alphabet);
        els.forEach(function (c, i) {
            var d = document.createElement("div");
            d.textContent = labs[i];
            d.style.cssText =
                "position:fixed;left:" + Math.max(0, c.rect.left - 2) + "px;top:" +
                Math.max(0, c.rect.top - 2) + "px;background:#f8e08e;color:#1a1a1a;" +
                "font:bold 11px monospace;padding:1px 4px;border-radius:3px;" +
                "box-shadow:0 1px 3px rgba(0,0,0,.55);";
            box.appendChild(d);
            items.push({ el: c.el, label: labs[i], div: d });
        });
        root.appendChild(box);
        document.documentElement.appendChild(host);
        return els.length;
    }

    function filter(typed) {
        var live = 0;
        items.forEach(function (it) {
            var m = it.label.indexOf(typed) === 0;
            it.div.style.display = m ? "" : "none";
            if (m) live++;
        });
        return live;
    }

    function activate(typed, newTab) {
        var it = null;
        for (var i = 0; i < items.length; i++)
            if (items[i].label === typed) { it = items[i]; break; }
        clear();
        if (!it) return { miss: true };
        var el = it.el;
        if (newTab && el.href) return { open: el.href };
        if (editable(el)) {
            el.focus();                     // editable.js flips us to insert
            return { focused: true };
        }
        el.focus();
        el.click();                         // full synthetic sequence, works on js-only buttons
        return { clicked: true };
    }

    function firstInput() {                 // gi
        var els = clickables();
        for (var i = 0; i < els.length; i++)
            if (editable(els[i].el)) { els[i].el.focus(); return true; }
        return false;
    }

    function clear() {
        if (host) host.remove();
        host = null;
        items = [];
    }

    return { show: show, filter: filter, activate: activate,
             firstInput: firstInput, clear: clear };
})();
